src/xdsl/error_format.py

```diff
@@ -0,0 +1,103 @@
+import os
+import sys
+from types import TracebackType
+
+
+class Colors:
+
+    class format:
+        reset = '\033[0m'
+        bold = '\033[01m'
+        underline = '\033[04m'
+
+    class fg:
+        red = '\033[31m'
+        green = '\033[32m'
+        orange = '\033[33m'
+        blue = '\033[34m'
+        cyan = '\033[36m'
+
+
+class Frame:
+    """
+    This class is a traceback wrapper of a given raised error 
+    """
+    e: Exception
+    extra_lines: int
```

Comment:
I am not sure how much information a code dump actually gives. In Python, installing packages basically works by copying code to your library folder. So if someone wants to look at the code, they could just open the respective file on the respective line number. Then, they can also scroll and look at more code above/below. I think this workflow makes more sense than having to specify the size of a code snippet ahead of time that will be dumped for an exception that occurs anywhere. 

Also, I feel that the most important thing someone can get from a code dump usually is the condition just above the `raise Exception`. And if someone has to rely on this if-clause in order to figure out what's wrong, we should write better error messages when raising exceptions instead of dumping code.

I think the most useful thing this functionality gives is a dump of `self`. This would for example makes [this PR](https://github.com/xdslproject/xdsl/pull/120) unneeded. In general, dumping the complete local context feels overkill to me, considering programs can get very complicated.

Reply:
That was a good call. I have thought about dumping just one line above the exception, but as a generic error/exception formatter that is based on `Diagnostic`, I would prefer it to have an argument that construct how many lines of code that would be in the dump, so in the future if we want to reconstruct the exception we can still use this to format it. Right now it's set to 3 as a placeholder.

The "how many lines of code is not an overkill" brings up another issue: to make sure that every raise exception has just enough information to be useful, each of the raise exceptions need to give its own context differently. (This requires quite a lot of work, preferably done incrementally). 

Such changes also affect `error_format.py`. In the future I would prefer this to be formatter only: diagnostic generates metadata and this module only formats it, so the traceback / exception is separated from the formatter. But right now I don't see this being done anytime soon.

Reply:
@webmiche errors may occurs at other places not inside a class function. So local variables may not contain ```self```. I'm still thinking better way to handle that. So far I can say we can reduce num of variables being dumped if it appears in previous frame.

***Side note***: We can even dump global variables but I don't think that's much use for it.

Reply:
I feel that the people look at these traces from bottom up, so it would probably be better to only dump things in their last context...

I think about the self problem, we might be able to just check whether self is bound to someting and then print it or not...

Reply:
Yes as it was mentioned here, i will try to make this possible. https://github.com/xdslproject/xdsl/pull/125#discussion_r902812027

Reply:
@webmiche I have changed the design so now the code snippet only shows the function that has the exception, can you please review it?